lib/Transforms/MemoryBanking.cpp

```diff
@@ -0,0 +1,408 @@
+//===- MemoryBanking.cpp - Code to perform memory bnaking in parallel loops
+//--------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements parallel loop memory banking.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/LLVM.h"
+#include "circt/Transforms/Passes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/LoopUtils.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/Dialect/Affine/Utils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace circt {
+#define GEN_PASS_DEF_MEMORYBANKING
+#include "circt/Transforms/Passes.h.inc"
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+
+namespace {
+
+/// Partition memories used in `affine.parallel` operation by the
+/// `bankingFactor` throughout the program.
+struct MemoryBankingPass
+    : public circt::impl::MemoryBankingBase<MemoryBankingPass> {
+  const std::function<unsigned(mlir::affine::AffineParallelOp)>
+      getBankingFactor;
+  MemoryBankingPass() : getBankingFactor(nullptr) {}
+  MemoryBankingPass(const MemoryBankingPass &other) = default;
+  explicit MemoryBankingPass(
+      std::optional<unsigned> bankingFactor = std::nullopt,
+      const std::function<unsigned(mlir::affine::AffineParallelOp)>
```

Comment:
Why is this function necessary? I'm not actually seeing where this is used

Reply:
It's not used indeed.
I referenced those affine loop transformation passes source code and saw them passing customized functions as arguments so that they can perform those customized transformation.
And I thought it could be useful in the future when the banking function is not as simple as taking the mod of the `bankingFactor` but some other complicated banking function (like TraceBank algorithm)
But I'm fine with removing it for now and introduce again in the future when it's actually used.

Please let me know what you think.

Reply:
Yes I'd say let's remove it for now, and then re-add it when we have a definitive reason. LGTM after this is done.